servicetalk-log4j2-mdc-utils/src/main/java/io/servicetalk/log4j2/mdc/utils/ServiceTalkThreadContextMap.java

```diff
@@ -42,6 +45,47 @@
     @SuppressWarnings("IllegalInstantiationOfString")
     private static final String NULL_STRING = "";
 
+    private static final String[] KNOWN_CONFLICTS = {
+            "io.servicetalk.log4j2.mdc.DefaultServiceTalkThreadContextMap",
```

Comment:
is this class used to differentiate between classes because `ServiceTalkTracingThreadContextMap` extends the class? Since the only conflict we are interested in involves one class extending another an alternative to adding additional classes and checking loaded classes is as follows:

```java
    private static final AtomicReference<String> loadedClassName = new AtomicReference<>();

    /**
     * Create a new instance.
     */
    public ServiceTalkThreadContextMap() {
        detectPossibleConflicts(getClass().getCanonicalName());
    }

    private static void detectPossibleConflicts(String thisClassName) {
        for (;;) {
            String existingClassName = loadedClassName.get();
            if (existingClassName == null) {
                if (loadedClassName.compareAndSet(null, thisClassName)) {
                    break;
                }
            } else if (existingClassName.equals(thisClassName)) {
                break;
            } else {
                System.err.println("Found conflicting MDC adapter " + existingClassName + " when attempting to load " +
                        thisClassName);
                // Log4j's provider interface provides no formal way to declare ordering or precedence if there are multiple
                // provides which may overlap. Currently log4j2's ThreadContextMapFactory gives precedent to the logger
                // specified via system property [1] so we check for this as a "best effort" to detect that this class has
                // precedence and will be used for MDC storage.
                // [1] https://logging.apache.org/log4j/2.0/log4j-api/apidocs/org/apache/logging/log4j/spi/ThreadContextMapFactory.html
                if (!System.getProperty("log4j2.threadContextMap", "").equals(thisClassName)) {
                    throw new IllegalStateException("log4j2.threadContextMap is not set to " +
                            thisClassName + ". There is no way to determine" +
                            " which ThreadContextMap has precedence for MDC storage and behavior is therefore undefined.");
                }
                // Do NOT use log4j2 to log here as it can cause some reentrancy problem
            }
        }
    }
```

Reply:
These classes won't be loaded together, so there is no need for the atomic operation. Actually, this solution won't report the issue. The problem lies to the order of the dependencies in your class path. If open tracing comes first, then that one will be loaded, otherwise the default mdc will. So there is no form of racing here. We just need to detect the presence of multiple behaviors in the CP and report to the user to avoid confusion; that's why we have to look for the classes specifically by name.

example:

- this will load the default solution
```
    implementation("io.servicetalk:servicetalk-log4j2-mdc:0.37.0-SNAPSHOT")
    implementation("io.servicetalk:servicetalk-opentracing-log4j2:0.37.0-SNAPSHOT")
```

- this will load the open tracing solution
```
    implementation("io.servicetalk:servicetalk-opentracing-log4j2:0.37.0-SNAPSHOT")
    implementation("io.servicetalk:servicetalk-log4j2-mdc:0.37.0-SNAPSHOT")
```

But not both.

Reply:
good point, implicit classpath loading too magical 😭 it feels wrong putting info on children classes in the parent class but I don't have a better idea (other than not doing anything)